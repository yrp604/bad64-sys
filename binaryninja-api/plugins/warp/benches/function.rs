use binaryninja::binary_view::BinaryViewExt;
use binaryninja::headless::Session;
use criterion::{criterion_group, criterion_main, Criterion};
use rayon::prelude::*;
use std::path::PathBuf;
use warp_ninja::build_function;

// These are the target files present in OUT_DIR
// Add the files to fixtures/bin
static TARGET_FILES: [&str; 1] = ["atox.obj"];

pub fn function_benchmark(c: &mut Criterion) {
    let session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    for file_name in TARGET_FILES {
        let path = out_dir.join(file_name);
        let view = session.load(&path).expect("Failed to load view");

        let functions: Vec<_> = view.functions().iter().map(|f| f.to_owned()).collect();
        // Bench all functions sequentially
        c.bench_function("signature all functions", |b| {
            b.iter(|| {
                for func in &functions {
                    let _ = build_function(&func, || func.lifted_il().ok(), false);
                }
            })
        });

        // Bench all functions in parallel
        c.bench_function("signature all functions rayon", |b| {
            b.iter(|| {
                functions
                    .par_iter()
                    .filter_map(|func| {
                        Some(build_function(
                            func.as_ref(),
                            || func.lifted_il().ok(),
                            false,
                        ))
                    })
                    .collect::<Vec<_>>()
            })
        });
    }
}

criterion_group!(benches, function_benchmark);
criterion_main!(benches);
