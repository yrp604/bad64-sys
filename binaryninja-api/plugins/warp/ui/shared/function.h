#pragma once
#include <QStandardItemModel>
#include <QTableView>

#include "binaryninjaapi.h"
#include "constraint.h"
#include "filter.h"
#include "misc.h"
#include "warp.h"

class WarpFunctionItem : public QStandardItem
{
    Warp::Ref<Warp::Function> m_function;

    // Optional attached data used to show/manage the function.
    Warp::Ref<Warp::Container> m_container;
    std::optional<Warp::Source> m_source;

public:
    WarpFunctionItem(Warp::Ref<Warp::Function> function,
                     BinaryNinja::Ref<BinaryNinja::Function> analysisFunction);

    void SetContainer(const Warp::Ref<Warp::Container> &container);

    void SetSource(Warp::Source source);

    Warp::Ref<Warp::Function> GetFunction() { return m_function; }
    Warp::Ref<Warp::Container> GetContainer() { return m_container; }
    std::optional<Warp::Source> GetSource() { return m_source; }
};

class WarpFunctionItemModel : public QStandardItemModel
{
    Q_OBJECT

    // The current matched function, used to highlight currently.
    Warp::Ref<Warp::Function> m_matchedFunction;

    // Mapping of function start address to the row index.
    // This is used to identify unique functions for updating instead of resetting the entire model.
    std::unordered_map<uint64_t, int> m_insertableFunctionRows;

public:
    WarpFunctionItemModel(const QStringList &labels, QObject *parent);

    static constexpr int COL_FUNCTION_ITEM = 0;
    static constexpr int COL_ADDRESS_ITEM = 1;

    void AppendFunction(WarpFunctionItem *item);

    void InsertFunction(uint64_t address, WarpFunctionItem *item);

    WarpFunctionItem *GetItem(const QModelIndex &index) const;

    std::optional<uint64_t> GetAddress(const QModelIndex &index) const;

    QVariant data(const QModelIndex &index, int role) const override;

    void SetMatchedFunction(const Warp::Ref<Warp::Function> &matchedFunction)
    {
        Warp::Ref<Warp::Function> previousMatchedFunction = m_matchedFunction;
        m_matchedFunction = matchedFunction;

        // Make sure to refresh the highlights so we don't keep the highlights from the previous function.
        if (previousMatchedFunction)
        {
            const QModelIndex topLeft = index(0, 0);
            const QModelIndex bottomRight = index(rowCount() - 1, 0);
            emit dataChanged(topLeft, bottomRight);
        }
    }
};

class WarpFunctionFilterModel : public QSortFilterProxyModel
{
    Q_OBJECT

public:
    WarpFunctionFilterModel(QObject *parent): QSortFilterProxyModel(parent)
    {
    }

    ~WarpFunctionFilterModel() override = default;

    bool filterAcceptsRow(int sourceRow, const QModelIndex &sourceParent) const override;

    bool lessThan(const QModelIndex &sourceLeft, const QModelIndex &sourceRight) const override;
};

class WarpFunctionTableWidget : public QWidget, public FilterTarget
{
    Q_OBJECT

    QTableView *m_table;
    WarpFunctionItemModel *m_model;
    WarpFunctionFilterModel *m_proxyModel;
    FilterEdit *m_filterEdit;
    FilteredView *m_filterView;
    QMenu *m_contextMenu;
    std::map<QString, std::function<void(WarpFunctionItem *, std::optional<uint64_t>)> > m_contextMenuActions;

public:
    explicit WarpFunctionTableWidget(QWidget *parent = nullptr);

    // TODO: Invert this and provide OnCallback functions that wrap the connect call.
    QTableView *GetTableView() const { return m_table; }
    WarpFunctionItemModel *GetModel() const { return m_model; }
    WarpFunctionFilterModel *GetProxyModel() const { return m_proxyModel; }

    void RegisterContextMenuAction(const QString &name,
                                   const std::function<void(WarpFunctionItem *, std::optional<uint64_t>)> &callback);

    void SetFunctions(QVector<WarpFunctionItem *> functions);

    void InsertFunction(uint64_t address, WarpFunctionItem *function);

    void setFilter(const std::string &) override;

    void scrollToFirstItem() override
    {
    }

    void scrollToCurrentItem() override
    {
    }

    void selectFirstItem() override
    {
    }

    void activateFirstItem() override
    {
    }
};

class WarpFunctionInfoWidget : public QWidget
{
    Q_OBJECT

    Warp::Ref<Warp::Function> m_function;
    BinaryNinja::Ref<BinaryNinja::Function> m_analysisFunction;

    // Optionally provide this information to show the source information.
    Warp::Ref<Warp::Container> m_container;
    std::string source;

    WarpConstraintTableWidget *m_constraintsTable;

    QTableView *m_commentsTable;
    QStandardItemModel *m_commentsModel;

    QTableView *m_variablesTable;

public:
    explicit WarpFunctionInfoWidget(QWidget *parent = nullptr);

    Warp::Ref<Warp::Function> GetFunction() { return m_function; }
    void SetFunction(Warp::Ref<Warp::Function> function) { m_function = function; };

    void SetAnalysisFunction(BinaryNinja::Ref<BinaryNinja::Function> analysisFunction)
    {
        m_analysisFunction = analysisFunction;
    };
    BinaryNinja::Ref<BinaryNinja::Function> GetAnalysisFunction() { return m_analysisFunction; }

    // TODO: Make this private?
    void UpdateInfo();
};
