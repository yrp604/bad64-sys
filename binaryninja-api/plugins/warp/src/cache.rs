pub mod container;
pub mod function;
pub mod guid;
pub mod type_reference;

pub use function::*;
pub use guid::*;
pub use type_reference::*;

use binaryninja::binary_view::{BinaryView, BinaryViewExt};
use binaryninja::function::Function as BNFunction;
use binaryninja::rc::Guard;
use binaryninja::rc::Ref as BNRef;
use binaryninja::ObjectDestructor;
use std::hash::{DefaultHasher, Hash, Hasher};

pub fn register_cache_destructor() {
    pub static mut CACHE_DESTRUCTOR: CacheDestructor = CacheDestructor;
    #[allow(static_mut_refs)]
    // SAFETY: This can be done as the backing data is an opaque ZST.
    unsafe {
        CACHE_DESTRUCTOR.register()
    };
}

/// A unique view ID, used for caching.
#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct ViewID(u64);

impl From<&BinaryView> for ViewID {
    fn from(value: &BinaryView) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write_u64(value.original_image_base());
        hasher.write_usize(value.file().session_id());
        Self(hasher.finish())
    }
}

impl From<BNRef<BinaryView>> for ViewID {
    fn from(value: BNRef<BinaryView>) -> Self {
        Self::from(value.as_ref())
    }
}

impl From<Guard<'_, BinaryView>> for ViewID {
    fn from(value: Guard<'_, BinaryView>) -> Self {
        Self::from(value.as_ref())
    }
}

/// A unique function ID, used for caching.
#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct FunctionID(u64);

impl From<&BNFunction> for FunctionID {
    fn from(value: &BNFunction) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write_u64(value.start());
        hasher.write_u64(value.lowest_address());
        hasher.write_u64(value.highest_address());
        Self(hasher.finish())
    }
}

impl From<BNRef<BNFunction>> for FunctionID {
    fn from(value: BNRef<BNFunction>) -> Self {
        Self::from(value.as_ref())
    }
}

impl From<Guard<'_, BNFunction>> for FunctionID {
    fn from(value: Guard<'_, BNFunction>) -> Self {
        Self::from(value.as_ref())
    }
}

pub struct CacheDestructor;

impl ObjectDestructor for CacheDestructor {
    fn destruct_view(&self, view: &BinaryView) {
        clear_type_ref_cache(view);
        log::debug!("Removed WARP caches for {:?}", view.file().filename());
    }
}
