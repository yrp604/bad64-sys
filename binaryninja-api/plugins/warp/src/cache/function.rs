use crate::convert::{comment_to_bn_comment, to_bn_symbol_at_address};
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::function::{Function as BNFunction, FunctionUpdateType};
use binaryninja::symbol::SymbolType;
use warp::signature::function::Function;

// TODO: Rename this?
/// Inserts a function match into the cache. This also has the side effect of setting persisted function
/// information, such as the matched function symbol.
///
/// IMPORTANT: This will mark the function as needing updates, if you intend to fill in functions with
/// no match (i.e. `None`), then you must change this function to prevent marking that as needing updates.
/// However, it's perfectly valid to remove a match and need to update the function still, so be careful.
pub fn insert_cached_function_match(function: &BNFunction, matched_function: Option<Function>) {
    let view = function.view();
    let function_start = function.start();
    // NOTE: If we expect to run match_function multiple times on a function, we should move this elsewhere.
    // Mark the function as needing updates so that reanalysis occurs on the function, and we apply the match.
    function.mark_updates_required(FunctionUpdateType::FullAutoFunctionUpdate);
    if let Some(auto_sym) = view.symbol_by_address(function_start) {
        // TODO: If we ever create non library function symbols we will need to remove this check (see: `to_bn_symbol_at_address`).
        if auto_sym.sym_type() == SymbolType::LibraryFunction {
            // NOTE: This will also mark for full auto function update, one thing to note is that the
            // requirement to call this is that this function not be called in the associated function's analysis.
            view.undefine_auto_symbol(&auto_sym);
        }
    }
    match matched_function {
        Some(matched_function) => {
            // Define the new matched function symbol, this can safely be done here as the symbol itself
            // will be persisted, unlike function type information or variable information.
            let new_sym = to_bn_symbol_at_address(&view, &matched_function.symbol, function_start);
            view.define_auto_symbol(&new_sym);
            // TODO: How to clear the comments? They are just persisted.
            // TODO: Also they generate an undo action, i hate implicit undo actions so much.
            for comment in &matched_function.comments {
                let bn_comment = comment_to_bn_comment(&function, comment.clone());
                function.set_comment_at(bn_comment.addr, &bn_comment.comment);
            }
            function.store_metadata("warp_matched_function", &matched_function.to_bytes(), false);
        }
        None => {
            function.remove_metadata("warp_matched_function");
        }
    }
}

// TODO: This does allocations, and for every reanalysis.
pub fn try_cached_function_match(function: &BNFunction) -> Option<Function> {
    let metadata = function.query_metadata("warp_matched_function")?;
    let raw_metadata = metadata.get_raw()?;
    Function::from_bytes(&raw_metadata)
}
