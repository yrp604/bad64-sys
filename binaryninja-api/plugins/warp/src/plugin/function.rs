use crate::cache::{cached_function_guid, try_cached_function_guid};
use crate::{get_warp_include_tag_type, INCLUDE_TAG_NAME};
use binaryninja::background_task::BackgroundTask;
use binaryninja::binary_view::{BinaryView, BinaryViewExt};
use binaryninja::command::{Command, FunctionCommand};
use binaryninja::function::Function;
use binaryninja::rc::Guard;
use rayon::iter::ParallelIterator;
use std::thread;
use warp::signature::function::FunctionGUID;

pub struct IncludeFunction;

impl FunctionCommand for IncludeFunction {
    fn action(&self, view: &BinaryView, func: &Function) {
        let sym_name = func.symbol().short_name();
        let sym_name_str = sym_name.to_string_lossy();
        let should_add_tag = func.function_tags(None, Some(INCLUDE_TAG_NAME)).is_empty();
        let insert_tag_type = get_warp_include_tag_type(view);
        match should_add_tag {
            true => {
                log::info!(
                    "Including selected function '{}' at 0x{:x}",
                    sym_name_str,
                    func.start()
                );
                func.add_tag(&insert_tag_type, "", None, false, None);
            }
            false => {
                log::info!(
                    "Removing included function '{}' at 0x{:x}",
                    sym_name_str,
                    func.start()
                );
                func.remove_tags_of_type(&insert_tag_type, None, false, None);
            }
        }
    }

    fn valid(&self, _view: &BinaryView, _func: &Function) -> bool {
        // TODO: Only allow if the function is named?
        true
    }
}

pub struct CopyFunctionGUID;

impl FunctionCommand for CopyFunctionGUID {
    fn action(&self, _view: &BinaryView, func: &Function) {
        let Some(guid) = cached_function_guid(func, || func.lifted_il().ok()) else {
            log::error!("Could not get guid for copied function");
            return;
        };
        log::info!(
            "Function GUID for {:?}... {}",
            func.symbol().short_name(),
            guid
        );
        if let Ok(mut clipboard) = arboard::Clipboard::new() {
            let _ = clipboard.set_text(guid.to_string());
        }
    }

    fn valid(&self, _view: &BinaryView, _func: &Function) -> bool {
        true
    }
}

pub struct FindFunctionFromGUID;

impl Command for FindFunctionFromGUID {
    fn action(&self, view: &BinaryView) {
        let Some(guid_str) = binaryninja::interaction::get_text_line_input(
            "Function GUID",
            "Find Function from GUID",
        ) else {
            return;
        };

        let Ok(searched_guid) = guid_str.parse::<FunctionGUID>() else {
            log::error!("Failed to parse function guid... {}", guid_str);
            return;
        };

        log::info!("Searching functions for GUID... {}", searched_guid);
        let funcs = view.functions();
        let view = view.to_owned();
        thread::spawn(move || {
            let background_task = BackgroundTask::new(
                &format!("Searching functions for GUID... {}", searched_guid),
                false,
            );

            // Only run this for functions which have already generated a GUID.
            let matched: Vec<Guard<Function>> = funcs
                .par_iter()
                .filter(|func| {
                    try_cached_function_guid(func).is_some_and(|guid| guid == searched_guid)
                })
                .collect();

            if matched.is_empty() {
                log::info!("No matches found for GUID... {}", searched_guid);
            } else {
                for func in &matched {
                    // Also navigate the user, as that is probably what they want.
                    if matched.len() == 1 {
                        let current_view = view.file().current_view();
                        if view
                            .file()
                            .navigate_to(&current_view, func.start())
                            .is_err()
                        {
                            log::error!(
                                "Failed to navigate to found function 0x{:0x} in view {}",
                                func.start(),
                                current_view
                            );
                        }
                    }
                    log::info!("Match found at function... 0x{:0x}", func.start());
                }
            }

            background_task.finish();
        });
    }

    fn valid(&self, _view: &BinaryView) -> bool {
        true
    }
}
