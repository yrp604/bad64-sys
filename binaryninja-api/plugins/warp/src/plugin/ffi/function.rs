use crate::build_function;
use crate::cache::{insert_cached_function_match, try_cached_function_match};
use crate::convert::{to_bn_symbol_at_address, to_bn_type};
use crate::plugin::ffi::{BNWARPConstraint, BNWARPFunction, BNWARPFunctionGUID};
use binaryninja::function::Function;
use binaryninja::rc::Ref;
use binaryninja::string::BnString;
use binaryninjacore_sys::{BNFunction, BNSymbol, BNType};
use std::ffi::c_char;
use std::mem::ManuallyDrop;
use std::sync::Arc;
use warp::signature::comment::FunctionComment;

#[repr(C)]
pub struct BNWarpFunctionComment {
    pub text: *mut c_char,
    pub offset: i64,
}

impl BNWarpFunctionComment {
    /// Leaks the text string to be freed with BNWARPFreeFunctionComment
    pub fn from_owned(value: &FunctionComment) -> Self {
        let text = BnString::into_raw(BnString::new(&value.text));
        Self {
            text,
            offset: value.offset,
        }
    }

    pub unsafe fn free_raw(value: &Self) {
        unsafe { BnString::free_raw(value.text) }
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPGetFunction(
    analysis_function: *mut BNFunction,
) -> *mut BNWARPFunction {
    let function = Function::from_raw(analysis_function);
    match build_function(&function, || function.lifted_il().ok(), false) {
        Some(function) => Arc::into_raw(Arc::new(function)) as *mut BNWARPFunction,
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPGetMatchedFunction(
    analysis_function: *mut BNFunction,
) -> *mut BNWARPFunction {
    let function = Function::from_raw(analysis_function);
    match try_cached_function_match(&function) {
        Some(matched_function) => {
            let arc_matched_function = Arc::new(matched_function);
            // NOTE: Freed by BNWARPFreeFunctionReference
            Arc::into_raw(arc_matched_function) as *mut BNWARPFunction
        }
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionApply(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) {
    let analysis_function = Function::from_raw(analysis_function);
    match function.is_null() {
        false => {
            // Set the matched function to `function`.
            let matched_function = ManuallyDrop::new(Arc::from_raw(function));
            insert_cached_function_match(
                &analysis_function,
                Some(matched_function.as_ref().clone()),
            )
        }
        true => {
            // We are removing the previous match.
            insert_cached_function_match(&analysis_function, None)
        }
    };
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetGUID(
    function: *mut BNWARPFunction,
) -> BNWARPFunctionGUID {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    function.guid
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetSymbol(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) -> *mut BNSymbol {
    let analysis_function = Function::from_raw(analysis_function);
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let view = analysis_function.view();
    let address = analysis_function.symbol().address();
    let function_symbol = to_bn_symbol_at_address(&view, &function.symbol, address);
    // NOTE: The symbol ref has been pre-incremented for the caller.
    Ref::into_raw(function_symbol).handle
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetSymbolName(function: *mut BNWARPFunction) -> *mut c_char {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let bn_name = BnString::new(&function.symbol.name);
    // NOTE: The symbol name string to be freed by BNFreeString
    BnString::into_raw(bn_name)
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetType(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) -> *mut BNType {
    let analysis_function = Function::from_raw(analysis_function);
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    match &function.ty {
        Some(func_ty) => {
            let arch = analysis_function.arch();
            let function_type = to_bn_type(Some(arch), func_ty);
            // NOTE: The type ref has been pre-incremented for the caller.
            unsafe { Ref::into_raw(function_type) }.handle
        }
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetConstraints(
    function: *mut BNWARPFunction,
    count: *mut usize,
) -> *mut BNWARPConstraint {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let raw_constraints: Box<[BNWARPConstraint]> = function
        .constraints
        .clone()
        .into_iter()
        .map(Into::into)
        .collect();
    *count = raw_constraints.len();
    let raw_constraints_ptr = Box::into_raw(raw_constraints);
    raw_constraints_ptr as *mut BNWARPConstraint
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetComments(
    function: *mut BNWARPFunction,
    count: *mut usize,
) -> *mut BNWarpFunctionComment {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let raw_comments: Box<[_]> = function
        .comments
        .iter()
        .map(BNWarpFunctionComment::from_owned)
        .collect();
    *count = raw_comments.len();
    let raw_comments_ptr = Box::into_raw(raw_comments);
    raw_comments_ptr as *mut BNWarpFunctionComment
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionsEqual(
    function_a: *mut BNWARPFunction,
    function_b: *mut BNWARPFunction,
) -> bool {
    // We do not own function so we should not drop.
    let function_a = ManuallyDrop::new(Arc::from_raw(function_a));
    // We do not own function so we should not drop.
    let function_b = ManuallyDrop::new(Arc::from_raw(function_b));
    function_a.eq(&function_b)
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPNewFunctionReference(
    function: *mut BNWARPFunction,
) -> *mut BNWARPFunction {
    Arc::increment_strong_count(function);
    function
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionReference(function: *mut BNWARPFunction) {
    if function.is_null() {
        return;
    }
    Arc::decrement_strong_count(function);
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionList(functions: *mut *mut BNWARPFunction, count: usize) {
    let functions_ptr = std::ptr::slice_from_raw_parts_mut(functions, count);
    let functions = Box::from_raw(functions_ptr);
    for function in functions {
        // NOTE: The functions themselves should also be arc.
        BNWARPFreeFunctionReference(function);
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionCommentList(
    comments: *mut BNWarpFunctionComment,
    count: usize,
) {
    let comments_ptr = std::ptr::slice_from_raw_parts_mut(comments, count);
    let comments = Box::from_raw(comments_ptr);
    for comment in &comments {
        BNWarpFunctionComment::free_raw(comment)
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeConstraintList(
    constraints: *mut BNWARPConstraint,
    count: usize,
) {
    let constraints_ptr = std::ptr::slice_from_raw_parts_mut(constraints, count);
    let _constraints = unsafe { Box::from_raw(constraints_ptr) };
}
