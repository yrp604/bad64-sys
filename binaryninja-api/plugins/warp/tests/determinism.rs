//! This tests to make sure the function GUIDs are stable.
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::headless::Session;
use std::collections::BTreeMap;
use std::path::PathBuf;
use warp::signature::function::FunctionGUID;
use warp_ninja::cache::cached_function_guid;

// These are the target files present in OUT_DIR
// Add the files to fixtures/bin
static TARGET_FILES: [&str; 10] = [
    "_ctype.obj",
    "_fptostr.obj",
    "_mbslen.obj",
    "_memicmp.obj",
    "_strnicm.obj",
    "_wctype.obj",
    "atof.obj",
    "atoldbl.obj",
    "atox.obj",
    "ls",
];

#[test]
fn insta_signatures() {
    let session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    for file_name in TARGET_FILES {
        let path = out_dir.join(file_name);
        let view = session.load(&path).expect("Failed to load view");
        let functions: BTreeMap<u64, FunctionGUID> = view
            .functions()
            .iter()
            .map(|f| {
                let guid = cached_function_guid(&f, || f.lifted_il().ok());
                (f.start(), guid.unwrap())
            })
            .collect();
        insta::assert_debug_snapshot!(file_name, functions);
    }
}
