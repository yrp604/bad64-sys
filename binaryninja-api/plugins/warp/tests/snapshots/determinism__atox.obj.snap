---
source: plugins/warp/tests/determinism.rs
expression: functions
---
{
    220944: FunctionGUID {
        guid: 9350843d-64a5-59a9-b39d-490b424fc116,
    },
    221264: FunctionGUID {
        guid: d04f0e79-31ec-5e17-83a5-9824ab6a3956,
    },
    221760: FunctionGUID {
        guid: 63d8d7e2-ffe2-5445-b68d-3d0863f98213,
    },
    222304: FunctionGUID {
        guid: fb5e8536-7b76-5b0d-ae32-2d9ab4834804,
    },
    222688: FunctionGUID {
        guid: fb5e8536-7b76-5b0d-ae32-2d9ab4834804,
    },
    223072: FunctionGUID {
        guid: d66e739b-41e2-55d4-848a-e2b095c31b88,
    },
    225616: FunctionGUID {
        guid: fe53c8f7-51c1-51c5-ad5f-8490ef43d23d,
    },
    228160: FunctionGUID {
        guid: 3038c695-7598-597c-b108-0bf3545bc4fa,
    },
    230912: FunctionGUID {
        guid: 9111b73b-0d09-57f7-8eab-fcdcdd16b931,
    },
    233680: FunctionGUID {
        guid: d4f8fff7-86f6-5f81-a80d-b1cf7b7e9035,
    },
    234128: FunctionGUID {
        guid: d4f8fff7-86f6-5f81-a80d-b1cf7b7e9035,
    },
    234592: FunctionGUID {
        guid: d4f8fff7-86f6-5f81-a80d-b1cf7b7e9035,
    },
    235056: FunctionGUID {
        guid: d4f8fff7-86f6-5f81-a80d-b1cf7b7e9035,
    },
    235520: FunctionGUID {
        guid: 25bc1f0b-4419-5e8b-9957-27a30009bcb1,
    },
    235952: FunctionGUID {
        guid: 25bc1f0b-4419-5e8b-9957-27a30009bcb1,
    },
    236384: FunctionGUID {
        guid: 8a0c3535-d4db-5c0e-93ec-13830ae546d2,
    },
    237072: FunctionGUID {
        guid: 3496f794-99ca-5c1a-b7b1-6adc6302b53c,
    },
    237456: FunctionGUID {
        guid: 3496f794-99ca-5c1a-b7b1-6adc6302b53c,
    },
    237840: FunctionGUID {
        guid: a8036a14-9afb-59fe-ba38-8102a56c0e0d,
    },
    238192: FunctionGUID {
        guid: 119c08fa-5fb3-529e-9bc8-feb53495aeef,
    },
    238496: FunctionGUID {
        guid: 730daa1c-47e1-5bf8-a17c-03151d3fa093,
    },
    238832: FunctionGUID {
        guid: 3e626237-c5e2-5175-914a-2b042e18362d,
    },
    239184: FunctionGUID {
        guid: 75f7c665-4972-5e97-9381-e81903e81932,
    },
    239536: FunctionGUID {
        guid: 463b6374-33ab-504b-8ca4-40d4707f17d9,
    },
    239904: FunctionGUID {
        guid: e137bdfa-4f15-54b4-ad68-b546a86a36e8,
    },
    240256: FunctionGUID {
        guid: 3fbc283b-91a6-53e4-8d6b-69b4e35b817b,
    },
    240560: FunctionGUID {
        guid: 73b5ec71-b4b8-5c40-9bf3-08e9cb6706c3,
    },
    240864: FunctionGUID {
        guid: 598a91b9-f84c-5fa8-a607-6b2d196009c4,
    },
    241168: FunctionGUID {
        guid: 6546d27d-6260-5a3b-b2b5-59a31979761f,
    },
    241472: FunctionGUID {
        guid: f61c75b2-0eb0-58a7-be62-1454cba9543a,
    },
    241904: FunctionGUID {
        guid: f6f56e01-894c-598b-bb38-1b85a3996830,
    },
    242384: FunctionGUID {
        guid: 3d2850d0-1e93-5e13-bd3d-7b285e3ff02f,
    },
    242752: FunctionGUID {
        guid: 3d2850d0-1e93-5e13-bd3d-7b285e3ff02f,
    },
    243120: FunctionGUID {
        guid: 960dbad0-a840-5d82-8403-605ede249f7a,
    },
    243456: FunctionGUID {
        guid: 960dbad0-a840-5d82-8403-605ede249f7a,
    },
    243792: FunctionGUID {
        guid: bbe5afb4-3123-5dd3-92f6-f2b70caada8a,
    },
    244448: FunctionGUID {
        guid: b01885ae-584a-515e-a437-c6713a6655a3,
    },
    245120: FunctionGUID {
        guid: 518aa60b-91dc-55c4-a873-36b3771eb187,
    },
    245744: FunctionGUID {
        guid: 518aa60b-91dc-55c4-a873-36b3771eb187,
    },
    246368: FunctionGUID {
        guid: 9f556a61-19d4-55eb-9c4c-6951af464b91,
    },
    247632: FunctionGUID {
        guid: f55996f1-dbb1-5cda-8eb1-5d39c23445be,
    },
    248000: FunctionGUID {
        guid: 24bb0d12-ee2d-5f3b-b7d6-cee173baf054,
    },
    248336: FunctionGUID {
        guid: 5ee44559-4428-56c3-890c-ab5c3cda675b,
    },
    248640: FunctionGUID {
        guid: 7abc0122-603e-5d74-b54e-f598cef3a637,
    },
    248976: FunctionGUID {
        guid: 7abc0122-603e-5d74-b54e-f598cef3a637,
    },
    249296: FunctionGUID {
        guid: 7abc0122-603e-5d74-b54e-f598cef3a637,
    },
    249616: FunctionGUID {
        guid: 7abc0122-603e-5d74-b54e-f598cef3a637,
    },
    249936: FunctionGUID {
        guid: f567b9a0-834b-5823-9a8a-727eb715a8b1,
    },
    250288: FunctionGUID {
        guid: 5bc83844-878e-59e1-8272-48daee1d19ff,
    },
    250720: FunctionGUID {
        guid: 5ee44559-4428-56c3-890c-ab5c3cda675b,
    },
    251024: FunctionGUID {
        guid: 5ee44559-4428-56c3-890c-ab5c3cda675b,
    },
    251328: FunctionGUID {
        guid: 7abc0122-603e-5d74-b54e-f598cef3a637,
    },
    251664: FunctionGUID {
        guid: 7abc0122-603e-5d74-b54e-f598cef3a637,
    },
    251984: FunctionGUID {
        guid: 5ee44559-4428-56c3-890c-ab5c3cda675b,
    },
    252288: FunctionGUID {
        guid: 7abc0122-603e-5d74-b54e-f598cef3a637,
    },
    252608: FunctionGUID {
        guid: 5ee44559-4428-56c3-890c-ab5c3cda675b,
    },
    252912: FunctionGUID {
        guid: 7abc0122-603e-5d74-b54e-f598cef3a637,
    },
    253232: FunctionGUID {
        guid: 5ee44559-4428-56c3-890c-ab5c3cda675b,
    },
    253536: FunctionGUID {
        guid: 5ee44559-4428-56c3-890c-ab5c3cda675b,
    },
    253840: FunctionGUID {
        guid: 5ee44559-4428-56c3-890c-ab5c3cda675b,
    },
}
